Some changes that helped me with wildcard routes

yb66 opened the pull request:
Hi there,

I needed something to transfer my blog and gave Export a look, but with all its wildcard routes it wasn't a good fit, so I tried hacking together something but it was a bit dodgy, so I ended up coming back to Export and modifying that to my needs. It appears to have worked so I thought I'd see if you're interested in merging in the changes.

The main aim of I did was to get `export!` to take a block *and* a list of paths that you may want visited. I also got it to loop over the paths using an external iterator. That gave it a bit more flexibility, and by doing that I could actually scan the pages for internal links and add them to the pages list, so every page on my site with a link gets a visit. That's just an example, but it can be used as it was originally too.

There are a few other things, but that's the main one. I also changed the error handling to be a bit softer and more flexible, which works, but I suspect could be improved upon.

Anyway, give it a look and let me know if you're interested or have any feedback. If you think it's worth a merge then I'll work on some docs for the README.

Kind Regards,
iain

j15e commented:
Hi. Thanks for sharing. Looks good to me. As you said updating the README + fixing the travis-ci build (your could remove old ruby versions  from `.travis.yml` and remove old tests) would make it complete.

I'll add a few code specific comments.

j15e commented on Gemfile:
```diff
@@ -1,14 +1,18 @@
 source 'https://rubygems.org'
 
 group :development do
-  gem 'debugger' unless RUBY_VERSION < '1.9'
```
Perfect, we should remove any support for old rubies (ex. < 2.0.0)

j15e commented on Gemfile:
```diff
@@ -1,14 +1,18 @@
 group :test do
   gem 'rack-test'
   gem 'rake'
-  gem 'test-unit'
```
Remove the `test/` folder with olds tests too.

j15e commented on lib/sinatra/export.rb:
```diff
@@ -2,97 +2,288 @@
+      # @yield [builder] Gives a Builder instance to the block (see Builder) that is called for every path visited.
+      def initialize(app, paths: nil, skips: nil, use_routes: nil, filters: [], error_handler: nil )
         @app = app
+        @use_routes = 
```
I am not familiar to this multi-line syntax, I think it would be more standard to have simple 1 liner 

```ruby
@use_routes = paths.nil? && use_routes.nil? ? true : use_routes
```

yb66 replied:
It's the compromise between the longer and more readable if/else or the one line but less readable ternary. I'm happy to change it though.

j15e commented on lib/sinatra/export.rb:
```diff
@@ -2,97 +2,288 @@
         end
+        @enums.push @paths.to_enum
+
+        catch(:no_more_paths) {
```
`catch(:no_more_paths) {` -> `catch(:no_more_paths) do` 

Usually using `{ }` syntax only for inline code block ex. `items.select { |item| item.red? }`

yb66 replied:
I make a distinction between blocks that will return by using `{}` and block that I can ignore the return with the `do` syntax. I don't remember where I got that idea, Russ Olsen maybe. Happy to change this though.

j15e commented on lib/sinatra/export.rb:
```diff
@@ -2,97 +2,288 @@
+        handle_error_dir_not_found!(dir) unless dir.exist? && dir.directory?
+
+        if @use_routes
+          @enums.push self.send( :route_paths ).to_enum
```
Not sure why you add extra space in some method calls but I would remove them if possible ex. `self.send(:route_paths).to_enum`

yb66 replied:
I find them more readable, but no worries, will do.

j15e commented on lib/sinatra/export.rb:
```diff
@@ -2,97 +2,288 @@
+        # @param [String] path
+        # @param [Integer] status The expected response status code. Anything different and the error handler is called. Defaults to 200.
+        # @return [Rack::MockResponse]
+        def get_path path, status=nil
```
`def get_path(path, status=nil)`

yb66 commented:
Hi,

I've made those changes. I updated the Travis config as best I thought, but I wasn't sure what to do with thte JRuby stuff as I'm not sure the keyword args are supported for that so commented them out.

Any problems just let me know.

Regards,
iain

yb66 commented:
Hej,

I've updated the README. While I was writing it I realised that having access to the processing via the block would be more helpful before the static file is written, so I did that and added a spec for it and an example in README. I made that the last commit so if you don't agree just drop it, but it appears to work fine.

iain

yb66 commented:
Ok, it seems I've broken the Travis checks by adding in Yard. I don't know what to do to fix that, I tried adding it as a development dependency to the gemspec (which I didn't want to touch) but that's not done it. Any ideas?

j15e commented:
You could add the `yard` gem to combined group test + dev so that it get installed on the CI : 

```ruby
# Gemfile
group :development, :test do
  gem 'yard'
end
```

Or make the yard optional in the Rakefile like so :

```ruby
# Rakefile
begin
  require 'yard'
  YARD::Rake::YardocTask.new do |t|
    t.files   = ['lib/**/*.rb']   # optional
    t.options = ['--any', '--extra', '--opts'] # optional
    t.stats_options = ['--list-undoc']         # optional
  end
rescue LoadError
end
```

yb66 commented:
Thanks, that really helped, but v2.0.x is still failing, which led me to the knowledge that [required keyword arguments weren't supported until v2.1](https://robots.thoughtbot.com/ruby-2-keyword-arguments#required-keyword-arguments).

I've added checks to those methods, so it should pass now, fingers crossed!

iain

j15e commented:
Ok great I just test you branch with my existing project, still work great.

I will add you to the repo contributor and grant you access on rubygems.org so you can release a new version. Please update the `CHANGELOG.md` accordingly before publishing and sign the gem using a key as describe in http://guides.rubygems.org/security/#building-gems and add your public cert to `certs/`

yb66 commented:
Thanks Jean-Philippe, I’ll do that this weekend.

iain

On 17 Dec 2015, at 18:35, Jean-Philippe Doyle <notifications@github.com> wrote:

> Ok great I just test you branch with my existing project, still work great.
> 
> I will add you to the repo contributor and grant you access on rubygems.org so you can release a new version. Please update the CHANGELOG.md accordingly before publishing and sign the gem using a key as describe in http://guides.rubygems.org/security/#building-gems and add your public cert to certs/
> 
> —
> Reply to this email directly or view it on GitHub.
>
